### ISSUE TITLE
Fail to pass the sanity check

### ISSUE THREAD

**Issue by mouendless** [yahoo/CaffeOnSpark#166]
my program run normaly with minist dataset under yarn-cluster mode, through the log can find the tasks finished successfully, but an exception was threw from code below.
my cluster has one namenode and two workers
I have tuned the spark.scheduler.maxRegisteredResourcesWaitingTime but still not sloving the problem.
I wonder what's the idea behind this check and what could cause failure of passing it.

the sanity check in CaffeOnSpark:

``` scala
// Sanity check
    val numExecutors: Int = sc.getExecutorMemoryStatus.size
    val numDriver: Int = if (sc.isLocal) 0 else 1
    if ((size + numDriver) != sc.getExecutorMemoryStatus.size) {
      log.error("Requested # of executors: " + clusterSize + " actual # of executors:" + (numExecutors - numDriver) +
        ". Please try to set --conf spark.scheduler.maxRegisteredResourcesWaitingTime with a large value (default 30s)")
      throw new IllegalStateException("actual number of executors is not as expected")
    }
```

the log was like this

```
16/10/11 14:33:49 INFO TaskSetManager: Starting task 0.0 in stage 2.0 (TID 4, longong-dsj-kv2.lg.dscc, partition 0,PROCESS_LOCAL, 2110 bytes)
16/10/11 14:33:49 INFO TaskSetManager: Starting task 1.0 in stage 2.0 (TID 5, longong-dsj-kv2.lg.dscc, partition 1,PROCESS_LOCAL, 2110 bytes)
16/10/11 14:33:49 INFO BlockManagerInfo: Added broadcast_2_piece0 in memory on longong-dsj-kv2.lg.dscc:54330 (size: 2.6 KB, free: 511.1 MB)
16/10/11 14:33:51 INFO TaskSetManager: Finished task 0.0 in stage 2.0 (TID 4) in 2231 ms on longong-dsj-kv2.lg.dscc (1/2)
16/10/11 14:33:51 INFO DAGScheduler: ResultStage 2 (collect at CaffeOnSpark.scala:123) finished in 2.232 s
16/10/11 14:33:51 INFO DAGScheduler: Job 2 finished: collect at CaffeOnSpark.scala:123, took 2.272706 s
16/10/11 14:33:51 INFO TaskSetManager: Finished task 1.0 in stage 2.0 (TID 5) in 2196 ms on longong-dsj-kv2.lg.dscc (2/2)
16/10/11 14:33:51 INFO YarnClusterScheduler: Removed TaskSet 2.0, whose tasks have all completed, from pool 
16/10/11 14:33:51 INFO CaffeOnSpark: rank = 0, address = ,longong-dsj-kv2.lg.dscc:48732, hostname = longong-dsj-kv2.lg.dscc
16/10/11 14:33:51 INFO CaffeOnSpark: rank = 1, address = longong-dsj-kv2.lg.dscc:55515,, hostname = longong-dsj-kv2.lg.dscc
16/10/11 14:33:51 ERROR CaffeOnSpark: Requested # of executors: 2 actual # of executors:1. Please try to set --conf spark.scheduler.maxRegisteredResourcesWaitingTime with a large value (default 30s)                                                                                                                                      
16/10/11 14:33:51 ERROR ApplicationMaster: User class threw exception: java.lang.IllegalStateException: actual number of executors is not as expected
java.lang.IllegalStateException: actual number of executors is not as expected
    at com.yahoo.ml.caffe.CaffeOnSpark.setupTraining(CaffeOnSpark.scala:132)
    at com.yahoo.ml.caffe.CaffeOnSpark.trainWithValidation(CaffeOnSpark.scala:276)
    at com.yahoo.ml.caffe.CaffeOnSpark$.main(CaffeOnSpark.scala:42)
    at com.yahoo.ml.caffe.CaffeOnSpark.main(CaffeOnSpark.scala)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:498)
    at org.apache.spark.deploy.yarn.ApplicationMaster$$anon$2.run(ApplicationMaster.scala:558)
16/10/11 14:33:51 INFO ApplicationMaster: Final app status: FAILED, exitCode: 15, (reason: User class threw exception: java.lang.IllegalStateException: actual number of executors is not as expected) 
16/10/11 14:33:51 INFO SparkContext: Invoking stop() from shutdown hook 
```

the submit script was:

```
export CAFFE_ON_SPARK=/home/spark/caffeOnSpark/CaffeOnSpark                                                                                                           

export SPARK_WORKER_INSTANCES=2
export DEVICES=1

#export LD_LIBRARY_PATH=${CAFFE_ON_SPARK}/caffe-public/distribute/lib:${CAFFE_ON_SPARK}/caffe-distri/distribute/lib
export LD_LIBRARY_PATH=${LD_LIBRARY_PATH}:/usr/local/cuda-7.0/lib64:/usr/local/mkl/lib/intel64/
export LD_LIBRARY_PATH=${LD_LIBRARY_PATH}:${CAFFE_ON_SPARK}/caffe-public/distribute/lib:${CAFFE_ON_SPARK}/caffe-distri/distribute/lib

hadoop fs -rm -f hdfs:///user/hadoop/caffe/mnist.model
hadoop fs -rm -r -f hdfs:///user/hadoop/caffe/mnist_features_result
spark-submit --master yarn --deploy-mode cluster \
        --num-executors ${SPARK_WORKER_INSTANCES} \
        --files ${CAFFE_ON_SPARK}/data/lenet_memory_solver.prototxt,${CAFFE_ON_SPARK}/data/lenet_memory_train_test.prototxt \
        --conf spark.driver.extraLibraryPath="${LD_LIBRARY_PATH}" \
        --conf spark.executorEnv.LD_LIBRARY_PATH="${LD_LIBRARY_PATH}" \
        --conf spark.scheduler.maxRegisteredResourcesWaitingTime="60s" \
        --class com.yahoo.ml.caffe.CaffeOnSpark  \
              ${CAFFE_ON_SPARK}/caffe-grid/target/caffe-grid-0.1-SNAPSHOT-jar-with-dependencies.jar \
        -train \
        -features accuracy,loss -label label \
        -conf lenet_memory_solver.prototxt \
        -devices ${DEVICES} \
        -connection ethernet \
        -model hdfs:///user/hadoop/caffe/mnist.model \
        -output hdfs:///user/hadoop/caffe/mnist_features_result
hadoop fs -ls hdfs:///user/hadoop/caffe/mnist.model
hadoop fs -cat hdfs:///user/hadoop/caffe/mnist_features_result/*
```

the prototxt file setting was

```
name: "LeNet"                                                                                                                                                         
layer {
  name: "data"
  type: "MemoryData"
  top: "data"
  top: "label"
  include {
    phase: TRAIN
  }
  source_class: "com.yahoo.ml.caffe.LMDB"
  memory_data_param {
    source: "hdfs:/projects/machine_learning/image_dataset/mnist_train_lmdb"
    batch_size: 64
    channels: 1
    height: 28
    width: 28
    share_in_parallel: false
  }
  transform_param {
    scale: 0.00390625
  }
}
layer {
  name: "data"
  type: "MemoryData"
  top: "data"
  top: "label"
  include {
    phase: TEST
  }
  source_class: "com.yahoo.ml.caffe.LMDB"
  memory_data_param {
    source: "hdfs:/projects/machine_learning/image_dataset/mnist_test_lmdb/"
    batch_size: 100 
    channels: 1
    height: 28
    width: 28
    share_in_parallel: false
  }
  transform_param {
    scale: 0.00390625
  }
}
..... //below are the same
```

**Comment by anfeng** [comment]
In YARN driver logs, do you see that 2 executors were launched?
The following entries in your log indicates that executor(s) are running on the same host:

- 16/10/11 14:33:51 INFO CaffeOnSpark: rank = 0, address = ,longong-dsj-kv2.lg.dscc:48732, hostname = longong-dsj-kv2.lg.dscc
- 16/10/11 14:33:51 INFO CaffeOnSpark: rank = 1, address = longong-dsj-kv2.lg.dscc:55515,, hostname = longong-dsj-kv2.lg.dscc

 
Likely, we just need to use additional Spark configurations.

**Comment by mouendless** [comment]
@anfeng yes, but in yarn-cluster node, one worker is selected as the driver, and the executors would prefere to  be signed on the others (still not include namenode), I think this is the spark framework strategy. And in the framework I think it's normal that one machine has more than one executor.

Besides, if I set the executor number to be 1 would get the error below.

```
16/10/11 10:51:31 INFO CaffeOnSpark: Schema:StructType(StructField(SampleID,StringType,false), StructField(accuracy,ArrayType(FloatType,true),false), StructField(loss,ArrayType(FloatType,true),false), StructField(label,ArrayType(FloatType,true),false))
16/10/11 10:51:31 INFO Utils: Fetching hdfs://blackpearl-00.dscc:11000/projects/machine_learning/image_dataset/mnist_test_lmdb/data.mdb to /home/hadoop/yarn/usercache/spark/appcache/application_1471244078305_0354/spark-143f9c3d-1d6b-4de1-a077-8507d476517d/userFiles-54bf570e-6565-462e-a195-a1a9da960462/mnist_test_lmdb/fetchFileTemp210134575770859623.tmp
16/10/11 10:51:32 INFO Utils: /home/hadoop/yarn/usercache/spark/appcache/application_1471244078305_0354/spark-143f9c3d-1d6b-4de1-a077-8507d476517d/userFiles-54bf570e-6565-462e-a195-a1a9da960462/mnist_test_lmdb/fetchFileTemp210134575770859623.tmp has been previously copied to /home/hadoop/yarn/usercache/spark/appcache/application_1471244078305_0354/spark-143f9c3d-1d6b-4de1-a077-8507d476517d/userFiles-54bf570e-6565-462e-a195-a1a9da960462/mnist_test_lmdb/data.mdb
16/10/11 10:51:32 INFO Utils: Fetching hdfs://blackpearl-00.dscc:11000/projects/machine_learning/image_dataset/mnist_test_lmdb/lock.mdb to /home/hadoop/yarn/usercache/spark/appcache/application_1471244078305_0354/spark-143f9c3d-1d6b-4de1-a077-8507d476517d/userFiles-54bf570e-6565-462e-a195-a1a9da960462/mnist_test_lmdb/fetchFileTemp7884231588866460403.tmp
16/10/11 10:51:32 ERROR ApplicationMaster: User class threw exception: org.apache.spark.sql.catalyst.errors.package$TreeNodeException: execute, tree:                 
TungstenAggregate(key=[], functions=[(count(1),mode=Final,isDistinct=false)], output=[count#36L])
+- TungstenExchange SinglePartition, None 
   +- TungstenAggregate(key=[], functions=[(count(1),mode=Partial,isDistinct=false)], output=[count#59L])
      +- InMemoryColumnarTableScan InMemoryRelation [SampleID#12,accuracy#13,loss#14,label#15], true, 10000, StorageLevel(true, false, false, false, 1), ConvertToUnsafe, None

org.apache.spark.sql.catalyst.errors.package$TreeNodeException: execute, tree:
TungstenAggregate(key=[], functions=[(count(1),mode=Final,isDistinct=false)], output=[count#36L])
+- TungstenExchange SinglePartition, None 
   +- TungstenAggregate(key=[], functions=[(count(1),mode=Partial,isDistinct=false)], output=[count#59L])
      +- InMemoryColumnarTableScan InMemoryRelation [SampleID#12,accuracy#13,loss#14,label#15], true, 10000, StorageLevel(true, false, false, false, 1), ConvertToUnsafe, None

    at org.apache.spark.sql.catalyst.errors.package$.attachTree(package.scala:49)
    at org.apache.spark.sql.execution.aggregate.TungstenAggregate.doExecute(TungstenAggregate.scala:80)
    at org.apache.spark.sql.execution.SparkPlan$$anonfun$execute$5.apply(SparkPlan.scala:132)
    at org.apache.spark.sql.execution.SparkPlan$$anonfun$execute$5.apply(SparkPlan.scala:130)
    at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:150)
    at org.apache.spark.sql.execution.SparkPlan.execute(SparkPlan.scala:130)
    at org.apache.spark.sql.execution.SparkPlan.executeCollect(SparkPlan.scala:166)
    at org.apache.spark.sql.execution.SparkPlan.executeCollectPublic(SparkPlan.scala:174)
    at org.apache.spark.sql.DataFrame$$anonfun$org$apache$spark$sql$DataFrame$$execute$1$1.apply(DataFrame.scala:1499)
    at org.apache.spark.sql.DataFrame$$anonfun$org$apache$spark$sql$DataFrame$$execute$1$1.apply(DataFrame.scala:1499)
    at org.apache.spark.sql.execution.SQLExecution$.withNewExecutionId(SQLExecution.scala:56)
    at org.apache.spark.sql.DataFrame.withNewExecutionId(DataFrame.scala:2086)
    at org.apache.spark.sql.DataFrame.org$apache$spark$sql$DataFrame$$execute$1(DataFrame.scala:1498)
    at org.apache.spark.sql.DataFrame.org$apache$spark$sql$DataFrame$$collect(DataFrame.scala:1505)
at org.apache.spark.sql.DataFrame$$anonfun$count$1.apply(DataFrame.scala:1515)
    at org.apache.spark.sql.DataFrame$$anonfun$count$1.apply(DataFrame.scala:1514)
    at org.apache.spark.sql.DataFrame.withCallback(DataFrame.scala:2099)
    at org.apache.spark.sql.DataFrame.count(DataFrame.scala:1514)
    at com.yahoo.ml.caffe.CaffeOnSpark.features(CaffeOnSpark.scala:432)
    at com.yahoo.ml.caffe.CaffeOnSpark$.main(CaffeOnSpark.scala:54)
    at com.yahoo.ml.caffe.CaffeOnSpark.main(CaffeOnSpark.scala)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:498)
    at org.apache.spark.deploy.yarn.ApplicationMaster$$anon$2.run(ApplicationMaster.scala:558)
Caused by: org.apache.spark.sql.catalyst.errors.package$TreeNodeException: execute, tree:
TungstenExchange SinglePartition, None
+- TungstenAggregate(key=[], functions=[(count(1),mode=Partial,isDistinct=false)], output=[count#59L])
   +- InMemoryColumnarTableScan InMemoryRelation [SampleID#12,accuracy#13,loss#14,label#15], true, 10000, StorageLevel(true, false, false, false, 1), ConvertToUnsafe, None

    at org.apache.spark.sql.catalyst.errors.package$.attachTree(package.scala:49)
    at org.apache.spark.sql.execution.Exchange.doExecute(Exchange.scala:247)
    at org.apache.spark.sql.execution.SparkPlan$$anonfun$execute$5.apply(SparkPlan.scala:132)
    at org.apache.spark.sql.execution.SparkPlan$$anonfun$execute$5.apply(SparkPlan.scala:130)
    at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:150)
    at org.apache.spark.sql.execution.SparkPlan.execute(SparkPlan.scala:130)
    at org.apache.spark.sql.execution.aggregate.TungstenAggregate$$anonfun$doExecute$1.apply(TungstenAggregate.scala:86)
    at org.apache.spark.sql.execution.aggregate.TungstenAggregate$$anonfun$doExecute$1.apply(TungstenAggregate.scala:80)
    at org.apache.spark.sql.catalyst.errors.package$.attachTree(package.scala:48)
    ... 25 more
Caused by: org.apache.spark.SparkException: File /home/hadoop/yarn/usercache/spark/appcache/application_1471244078305_0354/spark-143f9c3d-1d6b-4de1-a077-8507d476517d/userFiles-54bf570e-6565-462e-a195-a1a9da960462/mnist_test_lmdb/lock.mdb exists and does not match contents of hdfs://blackpearl-00.dscc:11000/projects/machine_learning/image_dataset/mnist_test_lmdb/lock.mdb
    at org.apache.spark.util.Utils$.copyFile(Utils.scala:490)
    at org.apache.spark.util.Utils$.downloadFile(Utils.scala:442)
    at org.apache.spark.util.Utils$.fetchHcfsFile(Utils.scala:625)
    at org.apache.spark.util.Utils$$anonfun$fetchHcfsFile$1.apply(Utils.scala:631)
    at org.apache.spark.util.Utils$$anonfun$fetchHcfsFile$1.apply(Utils.scala:630)
    at scala.collection.IndexedSeqOptimized$class.foreach(IndexedSeqOptimized.scala:33)
    at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:108)
    at org.apache.spark.util.Utils$.fetchHcfsFile(Utils.scala:630)
    at org.apache.spark.util.Utils$.doFetchFile(Utils.scala:600)
    at org.apache.spark.util.Utils$.fetchFile(Utils.scala:395)
    at org.apache.spark.SparkContext.addFile(SparkContext.scala:1386)
    at com.yahoo.ml.caffe.LmdbRDD.getPartitions(LmdbRDD.scala:45)
    at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:239)
    at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:237)
    at scala.Option.getOrElse(Option.scala:120)
    at org.apache.spark.rdd.RDD.partitions(RDD.scala:237)
    at org.apache.spark.rdd.MapPartitionsRDD.getPartitions(MapPartitionsRDD.scala:35)
    at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:239)
    at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:237)
    at scala.Option.getOrElse(Option.scala:120)
    at org.apache.spark.rdd.RDD.partitions(RDD.scala:237)
    at org.apache.spark.rdd.MapPartitionsRDD.getPartitions(MapPartitionsRDD.scala:35)
at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:239)
    at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:237)
    at scala.Option.getOrElse(Option.scala:120)
    at org.apache.spark.rdd.RDD.partitions(RDD.scala:237)
    at org.apache.spark.rdd.MapPartitionsRDD.getPartitions(MapPartitionsRDD.scala:35)
    at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:239)
    at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:237)
    at scala.Option.getOrElse(Option.scala:120)
    at org.apache.spark.rdd.RDD.partitions(RDD.scala:237)
    at org.apache.spark.rdd.MapPartitionsRDD.getPartitions(MapPartitionsRDD.scala:35)
    at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:239)
    at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:237)
    at scala.Option.getOrElse(Option.scala:120)
    at org.apache.spark.rdd.RDD.partitions(RDD.scala:237)
    at org.apache.spark.rdd.MapPartitionsRDD.getPartitions(MapPartitionsRDD.scala:35)
    at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:239)
    at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:237)
    at scala.Option.getOrElse(Option.scala:120)
    at org.apache.spark.rdd.RDD.partitions(RDD.scala:237)
    at org.apache.spark.rdd.MapPartitionsRDD.getPartitions(MapPartitionsRDD.scala:35)
    at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:239)
    at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:237)
    at scala.Option.getOrElse(Option.scala:120)
    at org.apache.spark.rdd.RDD.partitions(RDD.scala:237)
    at org.apache.spark.rdd.MapPartitionsRDD.getPartitions(MapPartitionsRDD.scala:35)
    at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:239)
    at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:237)
    at scala.Option.getOrElse(Option.scala:120)
    at org.apache.spark.rdd.RDD.partitions(RDD.scala:237)
    at org.apache.spark.rdd.MapPartitionsRDD.getPartitions(MapPartitionsRDD.scala:35)
    at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:239)
    at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:237)
    at scala.Option.getOrElse(Option.scala:120)
    at org.apache.spark.rdd.RDD.partitions(RDD.scala:237)
    at org.apache.spark.ShuffleDependency.<init>(Dependency.scala:91)
    at org.apache.spark.sql.execution.Exchange.prepareShuffleDependency(Exchange.scala:220)
    at org.apache.spark.sql.execution.Exchange$$anonfun$doExecute$1.apply(Exchange.scala:254)
    at org.apache.spark.sql.execution.Exchange$$anonfun$doExecute$1.apply(Exchange.scala:248)
    at org.apache.spark.sql.catalyst.errors.package$.attachTree(package.scala:48)
    ... 33 more
16/10/11 10:51:32 INFO ApplicationMaster: Final app status: FAILED, exitCode: 15, (reason: User class threw exception: org.apache.spark.sql.catalyst.errors.package$TreeNodeException: execute, tree:
TungstenAggregate(key=[], functions=[(count(1),mode=Final,isDistinct=false)], output=[count#36L])
+- TungstenExchange SinglePartition, None
   +- TungstenAggregate(key=[], functions=[(count(1),mode=Partial,isDistinct=false)], output=[count#59L])
      +- InMemoryColumnarTableScan InMemoryRelation [SampleID#12,accuracy#13,loss#14,label#15], true, 10000, StorageLevel(true, false, false, false, 1), ConvertToUnsafe, None
)
```

**Comment by anfeng** [comment]
Your log indicates that lock.lmdb is causing the problem. This issue was fixed by PR #155. Please get the latest CaffeOnSpark code.
